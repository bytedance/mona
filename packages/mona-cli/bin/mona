#! /usr/bin/env node

const mona = require('../dist').default;
mona();
