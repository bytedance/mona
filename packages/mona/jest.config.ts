import baseConfig from '../../jestconfig.base';

module.exports = baseConfig;
