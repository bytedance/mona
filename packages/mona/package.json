{
  "name": "@bytedance/mona",
  "version": "0.2.0-alpha.9",
  "description": "Merchant app's developing and building tools",
  "main": "./dist/index",
  "scripts": {
    "build": "yarn clean && ttsc -b",
    "start": "ttsc -b -w",
    "clean": "rm -rf dist",
    "test": "jest"
  },
  "files": [
    "dist"
  ],
  "author": "zhangxiaowei.chris@bytedance.com",
  "license": "MIT",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "devDependencies": {
    "copy-webpack-plugin": "^10.2.0",
    "react": "^17.0.2",
    "webpack": "^5.58.2",
    "webpack-chain": "^6.5.1"
  },
  "peerDependencies": {
    "react": "^17.0.2"
  },
  "gitHead": "cc5aa5283af595fd7d8a7b4921f54e2797504edd"
}
