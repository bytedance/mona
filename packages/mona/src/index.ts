export { createProjectConfig, ProjectConfig } from './createProjectConfig';
export { createAppConfig, AppConfig } from './createAppConfig';
export { createPageConfig, PageConfig } from './createPageConfig';
export { default as BaseApis } from './BaseApis';
export * from './BaseApis';
export { default as BaseComponents } from './BaseComponents';
export * from './BaseComponents';
export * from './types';
export * from './lifecycle';
