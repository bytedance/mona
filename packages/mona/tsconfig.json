{
  "compilerOptions": {
    "target": "es2017",
    "moduleResolution": "node",
    "strict": true,
    "module": "ESNext",
    "esModuleInterop": true,
    "resolveJsonModule": true,
    "allowSyntheticDefaultImports": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "forceConsistentCasingInFileNames": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "skipLibCheck": true,
    "pretty": true,
    "jsx": "react",
    "sourceMap": true,
    "declaration": true,
    "plugins": [
      {
        "transform": "typescript-transform-paths"
      },
      {
        "transform": "typescript-transform-paths",
        "afterDeclarations": true
      }
    ],
    "outDir": "dist",
    "baseUrl": ".",
    "paths": {
      "@/*": ["src/*"]
    }
  },
  "include": ["src"],
  "exclude": [
    "**/__tests__",
    "**/*.test.ts",
    "**/*.spec.ts",
    "**/*.test.tsx",
    "**/*.spec.tsx"
   ]
}
