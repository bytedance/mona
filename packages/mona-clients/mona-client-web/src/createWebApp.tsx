import { AppConfig } from '@bytedance/mona';
import React, { useEffect } from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter, Switch, Route, Redirect, useHistory } from 'react-router-dom';
import NavBar from './components/NavBar';
import TabBar from './components/TabBar';
import { formatPath, parseSearch, GLOBAL_LIFECYCLE_STORE } from '@bytedance/mona-shared';

export const WrapperComponent: React.FC<{ title: string }> = ({ children, title }) => {
  useEffect(() => {
    if (title) {
      document.title = title;
    } else {
      document.title = 'Mona Web';
    }
  }, [title]);

  return <>{children}</>;
};

const maskStyle: React.CSSProperties = {
  position: 'absolute',
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  zIndex: 1,
  backgroundColor: 'rgba(255, 255, 255)',
};

const NoMatchImgWrapperStyle: React.CSSProperties = {
  color: '#0000008C',
  fontSize: '14px',
  textAlign: 'center',
  lineHeight: '20px',
  whiteSpace: 'nowrap',
  position: 'absolute',
  left: '50%',
  top: '50%',
  transform: 'translate(-50%, -50%)',
};

const imgStyle = { width: 320, marginBottom: 8 };
export const NoMatch: React.FC<{ defaultPath: string }> = ({ defaultPath }) => {
  const history = useHistory();
  useEffect(() => {
    // app生命周期pageNotFound
    //@ts-ignore
    window[GLOBAL_LIFECYCLE_STORE]?.handlePageNotFound?.(defaultPath);
  }, [defaultPath]);
  return (
    <div style={maskStyle}>
      <div style={NoMatchImgWrapperStyle}>
        <img style={imgStyle} src={'https://lf3-fe.ecombdstatic.com/obj/ecom-open-butler/mona/error.png'} />
        <div>
          <span>
            不存在路由 {location.pathname} <a onClick={() => history.push(formatPath(defaultPath))}>返回首页</a>
          </span>
        </div>
      </div>
    </div>
  );
};

export interface PageProps {
  search: string;
  searchParams: Record<string, string>;
}

export const HistorySetWrapper: React.FC = ({ children }) => {
  const history = useHistory();

  // set global history to implement navigateTo and redirectTo api
  useEffect(() => {
    // @ts-ignore
    window.__mona_history = history;
  }, [history]);

  return <>{children}</>;
};

const defaultLightConfig: AppConfig['light'] = { mode: 'sidebar-semi-420' };

function prepareLightApp(config: AppConfig['light']) {
  // @ts-ignore
  if (typeof window.__MONA_LIGHT_APP_INIT_CB === 'function') {
    // @ts-ignore
    window.__MONA_LIGHT_APP_INIT_CB({ ...defaultLightConfig, ...config });
    // @ts-ignore
    window.__MONA_LIGHT_APP_INIT_CB = undefined;
  }
}

export function createWebApp(
  Component: React.ComponentType<any>,
  routes: { path: string; title: string; component: React.ComponentType<any> }[],
  options?: {
    tabBar?: AppConfig['tabBar'],
    navBar?: AppConfig['window'],
    defaultPath?: string
    light?: AppConfig['light']
  }
) {
  const render = ({ dom }: { dom: Element | Document }) => {
    prepareLightApp(options?.light)

    ReactDOM.render(
      <BrowserRouter>
        <HistorySetWrapper>
          <Component>
            {options?.navBar && <NavBar {...options?.navBar} />}
            <Switch>
              {routes?.map(route => (
                <Route
                  key={route.path}
                  path={formatPath(route.path)}
                  children={({ location }) => (
                    <WrapperComponent title={route.title}>
                      <route.component search={location.search} searchParams={parseSearch(location.search)} />
                    </WrapperComponent>
                  )}
                />
              ))}
              {routes?.length && (
                <Route exact path="/">
                  <Redirect to={formatPath(routes[0].path || options?.defaultPath || '/')} />
                </Route>
              )}
              <Route path="*">
                <NoMatch defaultPath={formatPath(routes[0].path || options?.defaultPath || '/')} />
              </Route>
            </Switch>
            {options?.tabBar && <TabBar tab={options?.tabBar} />}
          </Component>
        </HistorySetWrapper>
      </BrowserRouter>,
      dom.querySelector('#root'),
    );
  };

  // 在首次加载和执行时会触发该函数
  const provider = () => ({
    render, // 应用在渲染时会触发该 hook
    destroy({ dom }: { dom: Element }) {
      // 应用在销毁时会触发该 hook
      const root = (dom && dom.querySelector('#root')) || dom; // 若为 JS 入口直接将传入节点作为挂载点和销毁节点
      if (root) {
        // 做对应的销毁逻辑，保证子应用在销毁时对应的副作用也被移除
        ReactDOM.unmountComponentAtNode(root);
      }
    },
  });

  if (!window.__MARFISH__) {
    render({ dom: document });
  }

  return {
    provider,
  };
}
