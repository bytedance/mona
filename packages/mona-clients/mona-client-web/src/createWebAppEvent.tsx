import React from 'react';

import { LifecycleContext, AppLifecycleGlobalContext, AppLifecycle } from '@bytedance/mona';
import { isClassComponent, GLOBAL_LIFECYCLE_STORE } from '@bytedance/mona-shared';

export function createAppLifeCycle(Component: React.ComponentType<any>) {
  const appLifecycleContext = new LifecycleContext();
  const appEntryRef = React.createRef<any>();

  const callLifecycle = (callbackName: AppLifecycle, ...params: any[]) => {
    const cbs = appLifecycleContext.lifecycle[callbackName] || new Set([]);

    cbs.forEach(cb => cb(...params));

    if (appEntryRef.current?.[callbackName]) {
      return appEntryRef.current[callbackName](...params);
    }
  };

  const handleError = (...rest: any[]) => {
    callLifecycle(AppLifecycle.error, ...rest);
  };
  const handleLaunch = (...rest: any[]) => {
    callLifecycle(AppLifecycle.launch, ...rest);
  };
  const handleVisibilityChange = () => {
    if (document.visibilityState === 'visible') {
      callLifecycle(AppLifecycle.show);
    } else {
      callLifecycle(AppLifecycle.hide);
    }
  };
  const handlePageNotFound = (...rest: any[]) => {
    callLifecycle(AppLifecycle.pageNotFound, ...rest);
  };

  //@ts-ignore
  window[GLOBAL_LIFECYCLE_STORE] = {
    handleError,
    handleLaunch,
    handleVisibilityChange,
    handlePageNotFound,
  };

  class AppConfig extends React.Component {
    componentDidMount() {
      
      handleVisibilityChange();
      // onError
      window.addEventListener('error', handleError);
      // onShow & onHide
      document.addEventListener('visibilitychange', handleVisibilityChange);
      // onLaunch
      handleLaunch();
    }
    componentWillUnmount() {
      window.removeEventListener('error', handleError);
      document.removeEventListener('visibilitychange', handleVisibilityChange);
    }
    render() {
      if (isClassComponent(Component)) {
        return <Component {...this.props} ref={appEntryRef}></Component>;
      }
      return <Component {...this.props}></Component>;
    }
  }

  return (props: any) => (
    <AppLifecycleGlobalContext.Provider value={appLifecycleContext}>
      <AppConfig {...props} />
    </AppLifecycleGlobalContext.Provider>
  );
}
