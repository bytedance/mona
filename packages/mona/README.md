# @bytedance/mona
对外暴露提供构建时API
## 配置创建
### createProjectConfig
创建项目配置，`mona.config.ts`中使用
#### 例子
```js
// mona.config.ts
import { createProjectConfig } from '@bytedance/mona';
export default createProjectConfig({
  projectName: 'demo',
  input: './src/app.tsx',
  output: 'dist'
})
```

#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| projectName  | 应用/插件名称 | 是 | string | - |
| input  | 入口文件路径 | 否 | string | - |
| output  | 打包后的文件目录路径 | 否 | string | - |
| raw  | 定义函数来修改webpack配置 | 否 | (options: any) => any | - |
| dev  | 用来修改本地服务配置 | 否 | string | { port: 9999 } |

### createAppConfig
创建应用配置，`app.config.ts`中使用
#### 例子
```js
// app.config.ts
import { createAppConfig } from '@bytedance/mona';
export default createAppConfig({
  pages: [
    'pages/Home/index',
    'pages/Info/index',
    'pages/List/index'
  ]
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| pages  | 页面路径 | 是 | string[] | - |

默认`pages`的第一项为首页
### createPageConfig
创建页面配置，需在每个页面目录的`page.config.ts`中使用

#### 例子
```js
// page.config.ts
import { createPageConfig } from '@bytedance/mona';
export default createPageConfig({
  navigationBarTitleText: 'page title'
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| navigationBarTitleText  | 当前页面标题 | 否 | string | - |
