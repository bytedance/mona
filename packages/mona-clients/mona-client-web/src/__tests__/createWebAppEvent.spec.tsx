import React, { useEffect } from 'react';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { useAppEvent, AppLifecycle } from '@bytedance/mona';
import { createAppLifeCycle } from '../createWebAppEvent';
import enzyme, { configure, mount } from 'enzyme';
import mountTest from '../../../../tests/shared/mountTest';
import { NoMatch } from '../createWebApp';

configure({ adapter: new Adapter() });

describe('createAppLifecycle', () => {
  let launchMock = jest.fn();
  let PageNotFoundMock = jest.fn();
  let VisibleMock = jest.fn();
  let ErrorMock = jest.fn();

  beforeEach(() => {
    launchMock = jest.fn();
    PageNotFoundMock = jest.fn();
    VisibleMock = jest.fn();
    ErrorMock = jest.fn();
  });

  function App({ name }) {
    useAppEvent(AppLifecycle.launch, (...rest) => {
      launchMock(...rest, 'launch');
    });
    useAppEvent(AppLifecycle.pageNotFound, (...rest) => {
      PageNotFoundMock(...rest, 'pageNotFound');
    });
    useAppEvent(AppLifecycle.show, (...rest) => {
      VisibleMock(...rest, 'show');
    });
    useAppEvent(AppLifecycle.hide, (...rest) => {
      VisibleMock(...rest, 'hide');
    });
    // useAppEvent(AppLifecycle.error, (...rest) => {
    //   console.log('error出发了');
    // });
    return <>fc组件{name}</>;
  }
  class AppClass extends React.Component<{ name: string }> {
    onLaunch(...rest) {
      launchMock(...rest, 'onLoad');
    }
    onPageNotFound(...rest) {
      PageNotFoundMock(...rest, 'class', 'pageNotFound');
    }
    onShow(...rest) {
      VisibleMock(...rest, 'show');
    }
    onHide(...rest) {
      VisibleMock(...rest, 'hide');
    }
    render() {
      return <>class组件{this.props.name}</>;
    }
  }

  mountTest(createAppLifeCycle(App));
  mountTest(createAppLifeCycle(AppClass));

  it('createPageLifecycle FC lifeCycle', () => {
    const PageComponent = createAppLifeCycle(App);
    const AppIns = mount(<PageComponent name="mona" />);

    expect(AppIns.render()).toMatchSnapshot();

    // AppIns.props
    expect(AppIns.prop('name')).toBe('mona');

    expect(launchMock.mock.calls.length).toBe(1);
    AppIns.unmount();
  });
  it(`createPageLifecycle Class lifeCycle `, () => {
    const PageComponent = createAppLifeCycle(AppClass);
    const AppIns = mount(<PageComponent name="mona" />);
    // console.log("windowxxxx",window)
    // AppIns.instance.
    expect(AppIns.render()).toMatchSnapshot();

    expect(AppIns.prop('name')).toBe('mona');

    expect(launchMock.mock.calls.length).toBe(1);

    AppIns.unmount();
  });

  it('func pageNotFound', () => {
    const PageClassComponent = createAppLifeCycle(AppClass);
    mount(<PageClassComponent name="mona" />);
    mount(<NoMatch defaultPath="/path/a/b" />);
    // console.log(PageNotFoundMock.mock.calls);

    expect(PageNotFoundMock.mock.calls.length).toBe(1);
    expect(PageNotFoundMock.mock.calls).toMatchSnapshot();
    // PageNotFoundMock
  });
  it('class pageNotFound', () => {
    const PageComponent = createAppLifeCycle(App);
    mount(<PageComponent name="mona" />);
    mount(<NoMatch defaultPath="/path/a/b" />);
    // console.log(PageNotFoundMock.mock.calls);

    expect(PageNotFoundMock.mock.calls.length).toBe(1);
    expect(PageNotFoundMock.mock.calls).toMatchSnapshot();
    // PageNotFoundMock
  });

  it('func show & hide', () => {
    const PageComponent = createAppLifeCycle(App);

    mount(<PageComponent name="mona" />);
    expect(VisibleMock.mock.calls.length).toBe(1);
    expect(VisibleMock.mock.calls).toMatchSnapshot();
  });
  it('class show & hide', () => {
    const PageClassComponent = createAppLifeCycle(AppClass);

    mount(<PageClassComponent name="mona" />);
    expect(VisibleMock.mock.calls.length).toBe(1);
    expect(VisibleMock.mock.calls).toMatchSnapshot();
  });

  function AppError({ name }) {
    useAppEvent(AppLifecycle.error, (...rest) => {
      ErrorMock(...rest, 'error');
    });
    useEffect(() => {
      throw new Error('触发一个错误');
    }, []);
    return <>fc组件{name}</>;
  }
  class AppErrorClass extends React.Component<{ name: string; onClick: any }> {
    onError(...rest) {
      ErrorMock(...rest, 'error');
    }

    render() {
      return (
        <>
          class组件{this.props.name}
          <div className="error" onClick={this.props.onClick}></div>
        </>
      );
    }
  }

  it('func error', () => {
    const AppComponent = createAppLifeCycle(AppError);
    try {
      mount(<AppComponent name="mona" />);
    } catch {}
    expect(ErrorMock.mock.calls.length).toBe(1);
    expect(ErrorMock.mock.calls).toMatchSnapshot();
  });

  it('class error', () => {
    const AppComponent = createAppLifeCycle(AppErrorClass);
    try {
      const ins = mount(
        <AppComponent
          name="mona"
          onClick={() => {
            throw new Error('触发一个错误');
          }}
        />,
      );
      ins.find('.error').simulate('click', {});
    } catch {}
    expect(ErrorMock.mock.calls.length).toBe(1);
    expect(ErrorMock.mock.calls).toMatchSnapshot();
  });
});
