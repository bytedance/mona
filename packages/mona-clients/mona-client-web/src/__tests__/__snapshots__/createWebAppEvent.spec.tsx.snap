// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAppLifecycle class error 1`] = `
Array [
  Array [
    ErrorEvent {
      "isTrusted": true,
    },
    "error",
  ],
]
`;

exports[`createAppLifecycle class pageNotFound 1`] = `
Array [
  Array [
    "/path/a/b",
    "pageNotFound",
  ],
]
`;

exports[`createAppLifecycle class show & hide 1`] = `
Array [
  Array [
    "hide",
  ],
]
`;

exports[`createAppLifecycle createPageLifecycle Class lifeCycle  1`] = `
LoadedCheerio {
  "0": Node {
    "children": Array [
      Node {
        "data": "class组件mona",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "root",
    "next": null,
    "parent": null,
    "prev": null,
    "type": "root",
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "data": "class组件mona",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`createAppLifecycle createPageLifecycle FC lifeCycle 1`] = `
LoadedCheerio {
  "0": Node {
    "children": Array [
      Node {
        "data": "fc组件mona",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "root",
    "next": null,
    "parent": null,
    "prev": null,
    "type": "root",
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "data": "fc组件mona",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`createAppLifecycle func error 1`] = `
Array [
  Array [
    ErrorEvent {
      "isTrusted": true,
    },
    "error",
  ],
]
`;

exports[`createAppLifecycle func pageNotFound 1`] = `
Array [
  Array [
    "/path/a/b",
    "class",
    "pageNotFound",
  ],
]
`;

exports[`createAppLifecycle func show & hide 1`] = `
Array [
  Array [
    "hide",
  ],
]
`;
